### ISSUE TITLE
go_resource statements do not check out the specified revision to $GOPATH directory

### ISSUE THREAD

**Issue by maijs** [Homebrew/homebrew#44828]
I have a formula that has `go_resource` statements with URLs pointing to github.com repos with specific revisions.

    go_resource "labix.org/v2/mgo" do
      url "https://github.com/go-mgo/mgo.git",
        :revision => "2e26580ebcca4eca6533535f2dd062cd6bff44ac"
    end

When `Language::Go.stage_deps resources, buildpath/"src"` command stages the repos into `$GOPATH` folder, the repos are checked out to the tip of the tree instead of revision specified in the formula. This breaks the building of Go binaries.

Try running `brew install cayley -v` and you will see the breakage as the formula points to an older revision of `https://github.com/go-mgo/mgo.git` but it's getting the latest commit instead.

My setup: MacOS 10.11 with homebrewed `go-1.5.1`.

**Comment by maijs** [comment]
Output of `brew install cayley -v` is attached. You can see that the debug message is saying `Checking out revision [revision]` but after the message the actual `git checkout` command is never called. I added `sleep` in `def install` of `cayley.rb` to see which branches were checked out into `$GOPATH/src/github.com/[user]/[package]` and got confirmation that `$GOPATH/src/github.com/go-mgo/mgo` was at the latest commit, not `c6a7dce14133ccac2dcac3793f1d6e2ef048503a`.

This seems to be specific to `gitDownloadStrategy`, as `MercurialDownloadStrategy` is including revision hash into `archive` command: `/Users/mkr/bin/homebrew/opt/mercurial/bin/hg archive --subrepos -y -r 35bc42037350 -t files /private/tmp/cayley--code.google.com-p-go-uuid20151011-67465-1q5xpim`

    $ brew install cayley -v
    ==> Downloading https://github.com/google/cayley/archive/v0.4.1.tar.gz
    ==> Downloading from https://codeload.github.com/google/cayley/tar.gz/v0.4.1
    /usr/bin/curl -fLA Homebrew 0.9.5 (Ruby 2.0.0-645; OS X 10.11) https://codeload.github.com/google/cayley/tar.gz/v0.4.1 -C 0 -o /Users/mkr/Library/Caches/Homebrew/cayley-0.4.1.tar.gz.incomplete
      % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                     Dload  Upload   Total   Spent    Left  Speed
    100 4151k  100 4151k    0     0  1559k      0  0:00:02  0:00:02 --:--:-- 1558k
    ==> Verifying cayley-0.4.1.tar.gz checksum
    tar xf /Users/mkr/Library/Caches/Homebrew/cayley-0.4.1.tar.gz
    ==> Cloning https://github.com/badgerodon/peg.git
    git clone https://github.com/badgerodon/peg.git /Users/mkr/Library/Caches/Homebrew/cayley--github.com-badgerodon-peg--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--github.com-badgerodon-peg--git'...
    remote: Counting objects: 17, done.
    remote: Total 17 (delta 0), reused 0 (delta 0), pack-reused 17
    Unpacking objects: 100% (17/17), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision 9e5f7f4d07ca576562618c23e8abadda278b684f
    ==> Cloning https://github.com/barakmich/glog.git
    git clone https://github.com/barakmich/glog.git /Users/mkr/Library/Caches/Homebrew/cayley--github.com-barakmich-glog--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--github.com-barakmich-glog--git'...
    remote: Counting objects: 105, done.
    remote: Total 105 (delta 0), reused 0 (delta 0), pack-reused 105
    Receiving objects: 100% (105/105), 48.05 KiB | 0 bytes/s, done.
    Resolving deltas: 100% (56/56), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision fafcb6128a8a2e6360ff034091434d547397d54a
    ==> Cloning https://github.com/cznic/mathutil.git
    git clone https://github.com/cznic/mathutil.git /Users/mkr/Library/Caches/Homebrew/cayley--github.com-cznic-mathutil--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--github.com-cznic-mathutil--git'...
    remote: Counting objects: 339, done.
    remote: Total 339 (delta 0), reused 0 (delta 0), pack-reused 339
    Receiving objects: 100% (339/339), 223.75 KiB | 0 bytes/s, done.
    Resolving deltas: 100% (201/201), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision f9551431b78e71ee24939a1e9d8f49f43898b5cd
    ==> Cloning https://github.com/julienschmidt/httprouter.git
    git clone https://github.com/julienschmidt/httprouter.git /Users/mkr/Library/Caches/Homebrew/cayley--github.com-julienschmidt-httprouter--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--github.com-julienschmidt-httprouter--git'...
    remote: Counting objects: 508, done.
    remote: Compressing objects: 100% (3/3), done.
    remote: Total 508 (delta 0), reused 0 (delta 0), pack-reused 505
    Receiving objects: 100% (508/508), 196.23 KiB | 0 bytes/s, done.
    Resolving deltas: 100% (321/321), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision b59a38004596b696aca7aa2adccfa68760864d86
    ==> Cloning https://github.com/petar/GoLLRB.git
    git clone https://github.com/petar/GoLLRB.git /Users/mkr/Library/Caches/Homebrew/cayley--github.com-petar-GoLLRB-llrb--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--github.com-petar-GoLLRB-llrb--git'...
    remote: Counting objects: 175, done.
    remote: Total 175 (delta 0), reused 0 (delta 0), pack-reused 175
    Receiving objects: 100% (175/175), 9.76 MiB | 3.81 MiB/s, done.
    Resolving deltas: 100% (100/100), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision 53be0d36a84c2a886ca057d34b6aa4468df9ccb4
    ==> Cloning https://github.com/peterh/liner.git
    git clone https://github.com/peterh/liner.git /Users/mkr/Library/Caches/Homebrew/cayley--github.com-peterh-liner--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--github.com-peterh-liner--git'...
    remote: Counting objects: 444, done.
    remote: Compressing objects: 100% (13/13), done.
    remote: Total 444 (delta 5), reused 0 (delta 0), pack-reused 431
    Receiving objects: 100% (444/444), 129.80 KiB | 0 bytes/s, done.
    Resolving deltas: 100% (280/280), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision 1bb0d1c1a25ed393d8feb09bab039b2b1b1fbced
    ==> Cloning https://github.com/robertkrimen/otto.git
    git clone https://github.com/robertkrimen/otto.git /Users/mkr/Library/Caches/Homebrew/cayley--github.com-robertkrimen-otto--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--github.com-robertkrimen-otto--git'...
    remote: Counting objects: 2944, done.
    remote: Total 2944 (delta 0), reused 0 (delta 0), pack-reused 2944
    Receiving objects: 100% (2944/2944), 1.51 MiB | 1.23 MiB/s, done.
    Resolving deltas: 100% (1787/1787), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision d1b4d8ef0e0e4b088c8328c95ca63ab9ebd8fc9d
    ==> Cloning https://github.com/syndtr/gosnappy.git
    git clone https://github.com/syndtr/gosnappy.git /Users/mkr/Library/Caches/Homebrew/cayley--github.com-syndtr-gosnappy--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--github.com-syndtr-gosnappy--git'...
    remote: Counting objects: 119, done.
    remote: Compressing objects: 100% (44/44), done.
    remote: Total 119 (delta 69), reused 114 (delta 69), pack-reused 0
    Receiving objects: 100% (119/119), 44.00 KiB | 0 bytes/s, done.
    Resolving deltas: 100% (69/69), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision 156a073208e131d7d2e212cb749feae7c339e846
    ==> Cloning https://github.com/shurcooL/sanitized_anchor_name.git
    git clone https://github.com/shurcooL/sanitized_anchor_name.git /Users/mkr/Library/Caches/Homebrew/cayley--github.com-shurcooL-sanitized_anchor_name--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--github.com-shurcooL-sanitized_anchor_name--git'...
    remote: Counting objects: 29, done.
    remote: Total 29 (delta 0), reused 0 (delta 0), pack-reused 29
    Unpacking objects: 100% (29/29), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision 8e87604bec3c645a4eeaee97dfec9f25811ff20d
    ==> Cloning https://github.com/russross/blackfriday.git
    git clone https://github.com/russross/blackfriday.git /Users/mkr/Library/Caches/Homebrew/cayley--github.com-russross-blackfriday--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--github.com-russross-blackfriday--git'...
    remote: Counting objects: 1227, done.
    remote: Compressing objects: 100% (6/6), done.
    remote: Total 1227 (delta 0), reused 0 (delta 0), pack-reused 1221
    Receiving objects: 100% (1227/1227), 726.71 KiB | 856.00 KiB/s, done.
    Resolving deltas: 100% (698/698), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision 17bb7999de6cfb791d4f8986cc00b3309b370cdb
    ==> Cloning https://github.com/syndtr/goleveldb.git
    git clone https://github.com/syndtr/goleveldb.git /Users/mkr/Library/Caches/Homebrew/cayley--github.com-syndtr-goleveldb--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--github.com-syndtr-goleveldb--git'...
    remote: Counting objects: 4070, done.
    remote: Total 4070 (delta 0), reused 0 (delta 0), pack-reused 4070
    Receiving objects: 100% (4070/4070), 1.49 MiB | 1.25 MiB/s, done.
    Resolving deltas: 100% (2665/2665), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision 4875955338b0a434238a31165cb87255ab6e9e4a
    ==> Cloning https://github.com/boltdb/bolt.git
    git clone https://github.com/boltdb/bolt.git /Users/mkr/Library/Caches/Homebrew/cayley--github.com-boltdb-bolt--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--github.com-boltdb-bolt--git'...
    remote: Counting objects: 2396, done.
    remote: Compressing objects: 100% (4/4), done.
    remote: Total 2396 (delta 0), reused 0 (delta 0), pack-reused 2392
    Receiving objects: 100% (2396/2396), 992.54 KiB | 538.00 KiB/s, done.
    Resolving deltas: 100% (1566/1566), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision 3b449559cf34cbcc74460b59041a4399d3226e5a
    ==> Cloning https://github.com/go-mgo/mgo.git
    git clone https://github.com/go-mgo/mgo.git /Users/mkr/Library/Caches/Homebrew/cayley--gopkg.in-mgo.v2--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--gopkg.in-mgo.v2--git'...
    remote: Counting objects: 2611, done.
    remote: Total 2611 (delta 0), reused 0 (delta 0), pack-reused 2611
    Receiving objects: 100% (2611/2611), 5.64 MiB | 1.48 MiB/s, done.
    Resolving deltas: 100% (1795/1795), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision c6a7dce14133ccac2dcac3793f1d6e2ef048503a
    ==> Cloning https://github.com/go-mgo/mgo.git
    git clone https://github.com/go-mgo/mgo.git /Users/mkr/Library/Caches/Homebrew/cayley--gopkg.in-mgo.v2-unstable--git
    Cloning into '/Users/mkr/Library/Caches/Homebrew/cayley--gopkg.in-mgo.v2-unstable--git'...
    remote: Counting objects: 2611, done.
    remote: Total 2611 (delta 0), reused 0 (delta 0), pack-reused 2611
    Receiving objects: 100% (2611/2611), 5.64 MiB | 2.83 MiB/s, done.
    Resolving deltas: 100% (1795/1795), done.
    Checking connectivity... done.
    git config homebrew.cacheversion 0
    ==> Checking out revision 7a4943433e00707e38099a2b2e904d96681d14bc
    ==> Cloning https://code.google.com/p/go-uuid/
    /Users/mkr/bin/homebrew/opt/mercurial/bin/hg clone https://code.google.com/p/go-uuid/ /Users/mkr/Library/Caches/Homebrew/cayley--code.google.com-p-go-uuid--hg
    requesting all changes
    adding changesets
    adding manifests
    adding file changes
    added 15 changesets with 51 changes to 28 files
    updating to branch default
    17 files updated, 0 files merged, 0 files removed, 0 files unresolved
    ==> Checking out revision 35bc42037350
    /Users/mkr/bin/homebrew/opt/mercurial/bin/hg archive --subrepos -y -r 35bc42037350 -t files /private/tmp/cayley--code.google.com-p-go-uuid20151011-67465-1q5xpim
    ==> go build -o cayley
    src/github.com/syndtr/goleveldb/leveldb/table/reader.go:17:2: cannot find package "github.com/golang/snappy" in any of:
    	/Users/mkr/bin/homebrew/Cellar/go/1.5.1/libexec/src/github.com/golang/snappy (from $GOROOT)
    	/private/tmp/cayley20151011-67465-ekgnv7/cayley-0.4.1/src/github.com/golang/snappy (from $GOPATH)
    package .
    	imports github.com/google/cayley/internal: use of internal package not allowed
    package .
    	imports github.com/google/cayley/graph/mongo
    	imports gopkg.in/mgo.v2
    	imports gopkg.in/mgo.v2-unstable/internal/scram: use of internal package not allowed
    ==> Formula
    Path: /Users/mkr/bin/homebrew/Library/Formula/cayley.rb
    ==> Configuration
    HOMEBREW_VERSION: 0.9.5
    ORIGIN: https://github.com/Homebrew/Homebrew.git
    HEAD: 3c7289b325550e890f85e0ed8d43c2bfa2c615f3
    Last commit: 20 hours ago
    HOMEBREW_PREFIX: /Users/mkr/bin/homebrew
    HOMEBREW_REPOSITORY: /Users/mkr/bin/homebrew
    HOMEBREW_CELLAR: /Users/mkr/bin/homebrew/Cellar
    HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
    CPU: 8-core 64-bit haswell
    OS X: 10.11-x86_64
    Xcode: N/A
    CLT: 7.0.0.0.1.1441394355
    Clang: 7.0 build 700
    X11: N/A
    System Ruby: 2.0.0-p645
    Perl: /usr/bin/perl
    Python: /usr/bin/python
    Ruby: /usr/bin/ruby => /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby
    Java: 1.6.0_65-b14-468
    ==> ENV
    HOMEBREW_CC: clang
    HOMEBREW_CXX: clang++
    MAKEFLAGS: -j8
    CMAKE_PREFIX_PATH: /Users/mkr/bin/homebrew
    CMAKE_INCLUDE_PATH: /usr/include/libxml2:/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers
    CMAKE_LIBRARY_PATH: /System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries
    PKG_CONFIG_LIBDIR: /usr/lib/pkgconfig:/Users/mkr/bin/homebrew/Library/ENV/pkgconfig/10.11
    ACLOCAL_PATH: /Users/mkr/bin/homebrew/share/aclocal
    PATH: /Users/mkr/bin/homebrew/Library/ENV/4.3:/Users/mkr/bin/homebrew/opt/bazaar/bin:/Users/mkr/bin/homebrew/opt/go/bin:/Users/mkr/bin/homebrew/opt/mercurial/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Users/mkr/bin/homebrew/bin

    Error: cayley 0.4.1 did not build
    Logs:
         /Users/mkr/Library/Logs/Homebrew/cayley/01.go

    Warning: 
    ~ $

**Comment by apjanke** [comment]
Looks like the same issue as #44981. This will affect anything downloaded from Git, not just Go resources.

**Comment by apjanke** [comment]
I've put in a PR to fix this: #45019. With this, the `cayley` build still breaks, but it looks like it's getting the right resource revisions now, and failing for other reasons.

On OS X 10.9.5 with Xcode 6.2:

```
$ brew install --build-from-source cayley
==> Downloading https://github.com/google/cayley/archive/v0.4.1.tar.gz
==> Downloading from https://codeload.github.com/google/cayley/tar.gz/v0.4.1
...
==> go build -o cayley
Last 15 lines from /Users/janke/Library/Logs/Homebrew/cayley/01.go:
2015-10-16 04:04:46 -0400

go
build
-o
cayley

package .
	imports github.com/google/cayley/internal: use of internal package not allowed

READ THIS: https://git.io/brew-troubleshooting

[✘ /usr/local/Library/Formula on ⇄ git-fetch-revisions ±]
$
```

Could you try this PR out and see if it fixes your build issue?

**Comment by maijs** [comment]
@apjanke Thank you for your PR. It resolves the issue. My original problem was with #44884 for which I had to resort to tarball resource instead of git repo to get specific revision.
